### PR TITLE
Made committers extractable from WorkflowRun as well

### PR DESCRIPTION
**Pull request by erikhakansson**
Partially solves #277

Pipeline jobs are instances of WorkflowRun rather than AbstractBuild, thus retrieving changesets from a WorkflowRun is necessary to get changes.

However, Culprits are not yet available on WorkflowRun but I've included a preparation for when https://issues.jenkins-ci.org/browse/JENKINS-24141 gets solved.

### CODE REVIEW

**Comment by erikhakansson** [comment]
This PR is work in progress and will be rebased on #286 after that is merged.

**Comment by erikhakansson** [comment]
Since #286 was merged, I think this PR is ready for review

File: build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java [hunk 1]
```diff
@@ -144,36 +145,37 @@ public BuildViewModel previousBuild() {
...
-                        return entry.getAuthor().getFullName();
-                    }
-                });
-            }
-
-            private <T> T nonNullIterable(T list) {
-                return (T) getOrElse(list, newArrayList());
-            }
-        });
+        return getCommitters(build);
```
Comment (jan-molak) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: Could we extract this logic into a separate class? I'm worried that the BuildView is doing quite a few things already.

File: build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java [hunk 2]
```diff
@@ -208,19 +210,55 @@ private BuildView(Run<?, ?> build, boolean isPipeline, RelativeLocation parentJo
...
         if (build instanceof AbstractBuild<?, ?>) {
-            AbstractBuild<?, ?> jenkinsBuild = (AbstractBuild<?, ?>) build;
+            Iterable<String> iterable = transform(nonNullIterable(((AbstractBuild<?, ?>) build).getChangeSet()), new Function<ChangeLogSet.Entry, String>() {
+                @Override
+                public String apply(ChangeLogSet.Entry entry) {
+                    return entry.getAuthor().getFullName();
+                }
+            });
+            Iterables.addAll(committers, iterable);
+        } else if (build instanceof WorkflowRun) {
```
Comment (jan-molak) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: Looks like we have two cases here, so perhaps we could have two classes providing the list of committers? One for the regular builds, one for the workflows?

File: build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java [hunk 3]
```diff
@@ -208,19 +210,55 @@ private BuildView(Run<?, ?> build, boolean isPipeline, RelativeLocation parentJo
...
-            Iterables.addAll(users, reader.readUsersFrom(jenkinsBuild));
+                    @Override
+                    public String apply(@Nullable ChangeLogSet.Entry entry) {
+                        return entry != null ? entry.getAuthor().getFullName() : null;
+                    }
+                }));
+            }
         }
+        //If no committers were found, recursively get upstream committers:
+        if (committers.isEmpty()) {
```
Comment (jan-molak) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: And a third case here.

File: build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java (hunk 1 above)
Comment (ChadiEM) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: I've also done a similar logic in #304, in a class called `BuildStatusRetriever` (if WorkflowRun, do this, else do that). Perhaps the strategy pattern might help here: we can create a helper interface with two implementations, doing things differently whether a WorkflowRun or not. I would re-use that for the build status as well in #304 later on. What do you think, @jan-molak @erikhakansson?

Comment (jan-molak) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: Sounds like a good idea! @erikhakansson?

I think job-specific `Feature`s with a little sprinkling of the strategy pattern should make for quite a nice model.

Comment (erikhakansson) [inline build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java]: Yeah that seems like a good idea! I'll try to get some time to work on that!

**Comment by erikhakansson** [comment]
@jan-molak I've made some changes based on yours and @ChadiEM s suggestions. Please have a look!

**Comment by jan-molak** [comment]
Nice, I like the Retriever :-) But what's your thinking behind make the `StaticJenkinsAPIs` a singleton?

**Comment by erikhakansson** [comment]
@jan-molak hmm, I'm not sure. :) I could have sworn you asked me to but I must've dreamt that! I'll revert that part.

**Comment by erikhakansson** [comment]
@jan-molak had any time to look at the changes I did reverting the mysterious StaticJenkinsAPIs singletonification?

**Comment by jan-molak** [comment]
Hey @erikhakansson! Just tried to build the PR but and am getting a bunch of NPEs:

```
Results :

Tests in error: 
  HasHeadlineShowingFailedBuildDetailsTest.should_tell_the_number_of_broken_builds_since_the_last_build_broken_by_multiple_committers:49->headlineOf:92 » NullPointer
  HasHeadlineShowingFailedBuildDetailsTest.should_tell_the_number_of_broken_builds_since_the_last_broken_build_and_the_author_of_the_first_offending_commit:39->headlineOf:92 » NullPointer
  HasHeadlineShowingFailedBuildDetailsTest.should_tell_who_broke_the_build:20->headlineOf:92 » NullPointer
  HasHeadlineShowingFailedBuildDetailsTest.should_list_committers_who_broke_the_build_in_alphabetical_order:28->headlineOf:92 » NullPointer
  HasHeadlineShowingFixedBuildDetailsTest.should_list_committers_who_fixed_the_broken_build:31->headlineOf:80 » NullPointer
  HasHeadlineShowingFixedBuildDetailsTest.should_congratulate_anonymously_if_broken_build_was_fixed_without_known_committers:40->headlineOf:80 » NullPointer
  HasHeadlineShowingFixedBuildDetailsTest.should_tell_who_fixed_the_broken_build:22->headlineOf:80 » NullPointer
  HasHeadlineShowingRunningBuildDetailsTest.should_tell_whose_changes_are_being_built:29->headlineOf:60 » NullPointer
  HasHeadlineShowingRunningBuildDetailsTest.should_tell_whose_changes_are_being_built_when_there_are_multiple_committers:37->headlineOf:60 » NullPointer
  HasHeadlineShowingRunningBuildDetailsTest.should_say_nothing_if_no_builds_were_executed_and_one_is_running_now:21->headlineOf:60 » NullPointer

```

**Comment by erikhakansson** [comment]
Huh, weird I must've forgot to run the tests then. I'll have a look at it

**Comment by erikhakansson** [comment]
There, fixed now. I needed to mock the Jenkins instance to avoid that NPE. I didn't squash the commits so you can easily see what I did, in case you prefer it some other way :smile:

**Comment by erikhakansson** [comment]
By the way, ever considered using [travis-ci](https://travis-ci.org/) for build verification? It's free for open source projects and quite nice to for example instantly disqualify pull requests that causes NPEs in unit tests :laughing:

**Comment by jan-molak** [comment]
@erikhakansson - thanks! I set this project up a long time before Travis CI had such a nice integration with GitHub and decided to go with CloudBees Jenkins instead. I think Jenkins github plugin can be set up to build pull requests, but I didn't get the chance to look into this properly 😉

**Comment by jan-molak** [comment]
Looks like `mvn clean verify` fails with `NoClassDefFound` for `WorkflowRun` if Jenkins doesn't have the `workflow-aggregator` plugin installed.

**Comment by erikhakansson** [comment]
@jan-molak ah okay. I've only run the tests in intellij, I'll try building with maven and see if I can solve it.

**Comment by erikhakansson** [comment]
I'm having trouble running all the integration tests. We're behind a corporate proxy here, and the tests don't seem to respect my proxy settings. Any ideas on how I can solve that?

**Comment by erikhakansson** [comment]
I've committed what could possibly be a fix, but as I'm unable to run the integration tests I can't confirm. Could you please have a try?

Thanks

**Comment by xbournerx** [comment]
Is there an update to this PR?

**Comment by erikhakansson** [comment]
I'm not able to confirm the tests and @jan-molak seems to be away 😊

**Comment by erikhakansson** [comment]
@jan-molak Have you had any chance to look at this?

**Comment by erikhakansson** [comment]
@jan-molak any update on this?

**Comment by erikhakansson** [comment]
Rebased this on #370 and fixed acceptance tests.

**Comment by chrisallmark** [comment]
Is there an update to this PR?

**Comment by jan-molak** [comment]
#370 has now been merged in

**Comment by jan-molak** [comment]
Looks like there's a CloudBees<->BrowserStack connectivity problem on the CI server and the automated tests can't be run for this PR at the moment. I'm looking into it.

**Comment by jan-molak** [comment]
Issue with Jenkins resolved, but it looks like [the build for this PR](https://smartcode-opensource.ci.cloudbees.com/job/build-monitor/120/) is failing regardless so I had to revert it 😢

**Comment by dcendents** [comment]
It seems the problem is because the update center for stable-2.46 doesn't exist anymore.
i.e.: https://updates.jenkins-ci.org/stable-2.46/update-center.json
Not shown when you browse the directory:  https://updates.jenkins-ci.org/

But the http url works: http://updates.jenkins-ci.org/stable-2.46/update-center.json as it gets redirected to http://ftp-chi.osuosl.org/pub/jenkins/updates/stable-2.46/update-center.json (which contains all stable update sites since 1.532)

Simply changing `https` to `http` in `net.serenitybdd.integration.jenkins.environment.UpdateCenter` fixes the problem.

**Comment by jan-molak** [comment]
Awesome, thanks for looking into this! Would you mind creating a PR for this? I could merge it in today and then get this one in.

**Comment by dcendents** [comment]
@jan-molak I've added PR #376 which adds a working travis-ci configuration file and also fixes the current failing test (badge feature on jenkins 2.46.3, selenium issue really).

It would be great if you could merge it and enable travis-ci to build the project, then we could easily see if our pull requests build correctly and fix them.

You can see the result from my fork here: https://travis-ci.org/dcendents/jenkins-build-monitor-plugin/builds/391440045

**Comment by jan-molak** [comment]
Ha, using Travis CI would be sweet! Thanks so much for looking into this :-) Looks like CloudBees Jenkins is back to being happy again, so maybe let's get those PRs merged and look into migrating to Travis as the next step? 

One thing that CB Jenkins gives us that I'm not sure if Travis CI can is [publishing HTML reports from the acceptance tests](https://smartcode-opensource.ci.cloudbees.com/job/build-monitor/123/Serenity/). Would be good to figure out a way to do it should we want to migrate.

**Comment by dcendents** [comment]
Yes travis will not host reports, but you can always publish them on an external website. It does not even check the results of junit tests, it only checks if the command it ran returned with an error or not. So I would keep CB jenkins as well.

**Pull request closed without merging**